### PR TITLE
3djc/native chan mon

### PR DESCRIPTION
**Pull request by 3djc**
This would add Taranis like channel monitor. It also display true output instead of mixer output like the LUA counterpart does (we probably should remove the LUA version).

### CODE REVIEW

File: radio/src/gui/horus/widgets/outputs.cpp [hunk 1]
```diff
@@ -20,6 +20,10 @@
 
 #include "opentx.h"
 
+#define RECT_OFFSET 80
```
Comment (bsongis) [inline radio/src/gui/horus/widgets/outputs.cpp]: Perhaps align to 40 chars the 2 second column?

File: radio/src/gui/horus/widgets/outputs.cpp [hunk 2]
```diff
@@ -29,33 +33,60 @@ class OutputsWidget: public Widget
...
+    
+    uint8_t drawChannels(uint16_t x, uint16_t y, uint16_t w, uint16_t h, uint8_t firstChan)
+    {
+      char chanString[] = "CH32";
+      uint8_t lastChan = firstChan + floor(h / ROW_HEIGHT);
+      
+      for (uint8_t curChan = firstChan; curChan <= lastChan and curChan < 33; curChan++) {
+        int16_t chanVal = calcRESXto100(channelOutputs[curChan-1]);
+        
+        sprintf(chanString, "CH%02i", curChan);
```
Comment (bsongis) [inline radio/src/gui/horus/widgets/outputs.cpp]: avoid the ```sprintf``` here, rather use ```strAppendUnsigned```

File: radio/src/gui/horus/widgets/outputs.cpp [hunk 3]
```diff
@@ -29,33 +33,60 @@ class OutputsWidget: public Widget
...
+        lcdDrawText(x, y + (curChan - firstChan) * ROW_HEIGHT, chanString,SMLSIZE + TEXT_COLOR + LEFT);
+        strAppendSigned(chanString, chanVal);
```
Comment (bsongis) [inline radio/src/gui/horus/widgets/outputs.cpp]: You did well here :)

File: radio/src/gui/horus/widgets/outputs.cpp [hunk 4]
```diff
@@ -29,33 +33,60 @@ class OutputsWidget: public Widget
...
+        }
+        else if (chanVal < 0) {
+          uint16_t startpoint = x + RECT_OFFSET;
+          uint16_t endpoint = startpoint + floor(RECT_WIDTH / 2);
+          uint16_t size = RECT_WIDTH * abs(chanVal) / 200;
+          lcdDrawSolidFilledRect(endpoint - size,  y + (curChan - firstChan) * ROW_HEIGHT, size, ROW_HEIGHT, MAINVIEW_GRAPHICS_COLOR);
+        }
+      }
+      return lastChan;
+    };
```
Comment (bsongis) [inline radio/src/gui/horus/widgets/outputs.cpp]: no ```;``` here

File: radio/src/gui/horus/widgets/outputs.cpp [hunk 5]
```diff
@@ -29,33 +33,60 @@ class OutputsWidget: public Widget
...
+
+    void twoColumns()
+    {
+      uint8_t endColumn = drawChannels(zone.x, zone.y, floor(zone.w / 2), zone.h, persistentData->options[0].unsignedValue);
+      drawChannels(zone.x + floor(zone.w / 2) + 2, zone.y, floor(zone.w / 2), zone.h, endColumn + 1);
+    };
```
Comment (bsongis) [inline radio/src/gui/horus/widgets/outputs.cpp]: no ```;``` here

File: radio/src/gui/horus/widgets/outputs.cpp [hunk 6]
```diff
@@ -29,33 +33,60 @@ class OutputsWidget: public Widget
...
+    
+    void oneColumn()
+    {
+      drawChannels(zone.x, zone.y, zone.w, zone.h, persistentData->options[0].unsignedValue);
+    };
```
Comment (bsongis) [inline radio/src/gui/horus/widgets/outputs.cpp]: no ```;``` here

File: radio/src/gui/horus/widgets/outputs.cpp [hunk 7]
```diff
@@ -29,33 +33,60 @@ class OutputsWidget: public Widget
     }
 
     virtual void refresh();
+    
+    uint8_t drawChannels(uint16_t x, uint16_t y, uint16_t w, uint16_t h, uint8_t firstChan)
+    {
+      char chanString[] = "CH32";
+      uint8_t lastChan = firstChan + floor(h / ROW_HEIGHT);
```
Comment (bsongis) [inline radio/src/gui/horus/widgets/outputs.cpp]: Avoid working with floats whenever it's possible, here ```floor``` is not useful, the result will be always ```h / ROW_HEIGHT``` truncated.

**Comment by bsongis** [comment]
I see more usages of ```floor```
And I think you should use ```divRoundClosest``` instead of ```RECT_WIDTH * chanVal / 200```

**Comment by 3djc** [comment]
Git issue, yes , again

**Pull request closed without merging**